### PR TITLE
feat: enable customizing autoform layout

### LINKED ISSUE #1423: [autocrud] Provide your own form layout
It should be possible to use a custom layout for the form so you can make the usability better for the end user

### PR DESCRIPTION
**Pull request by taefi**
## Description

This changes the default layout of `<AutoForm />` from `VerticalLayout` to `FormLayout`, this means, if the `AutoForm` is used without any configurations regarding the layout (e.g. as `<AutoForm service={...} model={...} />`), it will render the fields using a `<FormLayout />` with the default two-column layout and the default `responsiveSteps`.

Also, this enables two levels of layout customizations via introducing `customLayoutRenderer` property that can be either of type:
```tsx
type AutoFormLayoutProps = Readonly<{
  template: FieldColSpan[][] | string[][];
  responsiveSteps?: Array<{ minWidth: string; columns: number }>;
}>;
```
which helps the developer customizing the underlying `FormLayout`,
Or 
```tsx
export type AutoFormLayoutRendererProps<M extends AbstractModel> = Readonly<{
  form: UseFormResult<M>;
  children: ReadonlyArray<ReactElement<AutoFormFieldProps>>;
}>;
```
that enables the developer to provide a function that accepts the form binder instance and the list of already bound fields, to create any custom layout component they desire.

Some examples of how this property can be used:

1. By defining a template for arranging the fields in the FormLayout, in either of the following shapes:

    - defining the template as `string[][]`:
        ```tsx
        <AutoForm service={service} model={PersonModel}
                        customLayoutRenderer={{
                            template: [
                                ['firstName', 'lastName', 'email'],
                                ['city', 'zipCode'],
                             ]
                        }}
        />
        ```
        This creates a FormLayout with 6 columns that renders the fields in two rows accordingly, the first three fields spanning over 2 columns, and the last two spanning over 3 columns.

    - defining the template as `{ property: string, colSpan: number }[][]`:
        ```tsx
        <AutoForm service={service} model={PersonModel}
                        customLayoutRenderer={{
                            template: [
                                [
                                    { property: 'firstName', colSpan: 2 },
                                    { property: 'lastName', colSpan: 2 },
                                    { property: 'email', colSpan: 2 },
                                ],
                                [
                                    { property: 'city', colSpan: 4 },
                                    { property: 'zipCode', colSpan: 2 },
                                ]
                            ]
                        }}
        />
        ```
        The above creates a 6 columns FormLayout (due having three fields in the first row of the template, and having two on the second row) that renders the fields in two rows, in which the first row's space is divided equally between those three fields, and in the second row, the `city` field takes twice the size as `zipCode` takes.

    - In addition to the template, it is also possible to configure the `responsiveSteps` for the form layout to have full control over the `FormLayout` behaviour:
        ```tsx
        <AutoForm service={service} model={PersonModel}
                        customLayoutRenderer={{
                            template: [
                                [
                                    { property: 'firstName', colSpan: 1 },
                                    { property: 'lastName', colSpan: 1 },
                                    { property: 'email', colSpan: 1 },
                                ],
                                [
                                    { property: 'city', colSpan: 2 },
                                    { property: 'zipCode', colSpan: 1 },
                                ]
                            ],
                            responsiveSteps: [
                                { minWidth: '0', columns: 1 },
                                { minWidth: '800px', columns: 3 }
                            ]
                        }}
        />
        ```
        The above creates a 3 columns FormLayout when the viewport is wider that 800 pixels, and renders the fields in two rows, in which the first row's space is divided equally between those three fields, and in the second row, the `city` field takes twice the size as `zipCode` takes. 

2. By providing a component to introduce any custom layout for the form: 

  ```tsx
  function MyVerticalLayoutRenderer({ children, form }: AutoFormLayoutRendererProps<PersonModel>) {
      return <VerticalLayout>{children}</VerticalLayout>;
  }

  <AutoForm service={service} model={PersonModel}
                     customLayoutRenderer={MyVerticalLayoutRenderer}
  />
  ```
  The above replaces the `FormLayout` with the component returned by the `MyVerticalLayoutRenderer` which in this example is simply a `VerticalLayout`. Note that the `children` are the read-only array of already bound fields to the `form` and there is no need to bound them here again. The `form` instance is passed to this function to provide more control to the developer e.g. for defining record level validations, defining custom fields and binding them, and etc.
    
Fixes #1423

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: packages/ts/react-crud/src/autoform.tsx [hunk 1]
```diff
@@ -17,23 +17,118 @@ type SubmitEvent<TItem> = {
...
+  customFormLayout: AutoFormLayoutProps,
+): Array<{ minWidth: string; columns: number }> {
+  function gcd(a: number, b: number): number {
+    return a > 0 ? gcd(b % a, a) : b;
+  }
+  const lcm = (a: number, b: number) => (a * b) / gcd(a, b);
+
+  const minNeededColumns = customFormLayout.template
+    .map((row) => row.length)
+    .filter((value, index, array) => array.indexOf(value) === index)
```
Comment (cromoteca) [inline packages/ts/react-crud/src/autoform.tsx:78]: Is this filter necessary?

File: packages/ts/react-crud/src/autoform.tsx [hunk 2]
```diff
@@ -17,23 +17,118 @@ type SubmitEvent<TItem> = {
...
+    });
+  } else {
+    weightedTemplate = customFormLayout.template as FieldColSpan[][];
+  }
+
+  const spannedFields: JSX.Element[] = [];
+  weightedTemplate.forEach((row: FieldColSpan[]) => {
+    row.forEach((fieldColSpan: FieldColSpan) => {
+      const field = fieldsByPropertyName.get(fieldColSpan.property)!;
+      const spannedField = React.cloneElement(field, { colspan: fieldColSpan.colSpan });
```
Comment (cromoteca) [inline packages/ts/react-crud/src/autoform.tsx:115]: This seems to be discouraged. Did you try one of the [alternatives](https://react.dev/reference/react/cloneElement#alternatives)?

File: packages/ts/react-crud/src/autoform.tsx [hunk 3]
```diff
@@ -72,16 +167,29 @@ export function ExperimentalAutoForm<TItem>({
...
-        {getProperties(model)
-          .filter(includeProperty)
-          .map((propertyInfo) => (
-            <AutoFormField key={propertyInfo.name} propertyInfo={propertyInfo} form={form} disabled={disabled} />
-          ))}
-        {formError ? <div style={{ color: 'var(--lumo-error-color)' }}>{formError}</div> : <></>}
-      </VerticalLayout>
+    <section className="flex flex-col p-m gap-m">
+      {layout}
+      {formError ? <div style={{ color: 'var(--lumo-error-color)' }}>{formError}</div> : <></>}
```
[the remark this answers is not in the document]
Comment (taefi) [reply packages/ts/react-crud/src/autoform.tsx:192]: I guess this form error is associated with the errors that happens while submitting and I guess there shouldn't be a problem of having them outside of the FormLayout component.

[the remark this answers is not in the document]
Comment (cromoteca) [reply packages/ts/react-crud/src/autoform.tsx:192]: There were two vertical layouts before. Maybe you could replace the inner one with your new layout, but leave the outer one to force errors and buttons below the form. Although I haven't tested that, I'm just looking at the code.

File: packages/ts/react-crud/src/autoform.tsx [hunk 4]
```diff
@@ -17,23 +17,118 @@ type SubmitEvent<TItem> = {
...
+type CustomFormLayoutProps = Readonly<{
+  customFormLayout: AutoFormLayoutProps;
+  children: ReadonlyArray<ReactElement<AutoFormFieldProps>>;
 }>;
 
-export function ExperimentalAutoForm<TItem>({
+function findColumnCount(responsiveSteps: Array<{ minWidth: string; columns: number }>): number {
+  return responsiveSteps
+    .map((step: { minWidth: string; columns: number }) => ({
+      minWidth: parseInt(step.minWidth, 10),
```
[the remark this answers is not in the document]
Comment (taefi) [reply packages/ts/react-crud/src/autoform.tsx:53]: I believe supporting a mixture of units would be an overkill (at least for the first version).

File: packages/ts/react-crud/src/autoform.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (taefi) [reply packages/ts/react-crud/src/autoform.tsx:192]: I think a section with flex direction of `flex-col` should render its children vertically same as `VerticalLayout.

File: packages/ts/react-crud/src/autoform.tsx [hunk 5]
```diff
@@ -17,23 +17,118 @@ type SubmitEvent<TItem> = {
...
+      columns: step.columns,
+    }))
+    .filter(({ minWidth }) => minWidth <= window.innerWidth)
+    .reduce(
+      (maxStep, step) => {
+        if (step.minWidth > maxStep.minWidth) {
+          return step;
+        }
+        return maxStep;
```
[the remark this answers is not in the document]
Comment (taefi) [reply packages/ts/react-crud/src/autoform.tsx:62]: Not sure how. `step` and `maxStep` are objects, and I need to reduce the stream to find the object with the biggest`minWidth` to get the `columns` from that.

File: packages/ts/react-crud/src/autoform.tsx (hunk 1 above)
Comment (taefi) [reply packages/ts/react-crud/src/autoform.tsx:78]: No, removed.

File: packages/ts/react-crud/src/autoform.tsx (hunk 2 above)
Comment (taefi) [reply packages/ts/react-crud/src/autoform.tsx]: Done.

File: packages/ts/react-crud/src/autoform.tsx [hunk 6]
```diff
@@ -17,23 +17,118 @@ type SubmitEvent<TItem> = {
...
-export type AutoFormProps<TItem> = Readonly<{
-  service: CrudService<TItem>;
-  model: DetachedModelConstructor<AbstractModel<TItem>>;
-  item?: TItem | typeof emptyItem;
+export type AutoFormLayoutRendererProps<M extends AbstractModel> = Readonly<{
+  form: UseFormResult<M>;
+  children: ReadonlyArray<ReactElement<AutoFormFieldProps>>;
+}>;
+
+interface FieldColSpan {
```
[the remark this answers is not in the document]
Comment (taefi) [reply packages/ts/react-crud/src/autoform.tsx]: Changed to type.

File: packages/ts/react-crud/src/autoform.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (cromoteca) [reply packages/ts/react-crud/src/autoform.tsx:62]: Sorry, my bad

[the remark this answers is not in the document]
Comment (taefi) [reply packages/ts/react-crud/src/autoform.tsx:62]: No worries.

File: packages/ts/react-crud/src/autoform.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (cromoteca) [reply packages/ts/react-crud/src/autoform.tsx:53]: Definitely. Also, I think it's quite common to use the same unit for all fields. But it should be pointed out in the docs.

**Comment by platosha** [comment]
For me, this seems to pack two properties into one with `customLayoutRenderer` supporting both props and renderer use cases. The type definition and the API doc description of such properties is convoluted. Another potentially confusing thing here is that a property named `smthRenderer` accepts something which is not a renderer.

I would suggest to separate them, so that we have two properties with distinct value shapes / use cases. E. g., `layout` for the properties, and `layoutRenderer` for the custom component.

**Comment by taefi** [comment]
> For me, this seems to pack two properties into one with `customLayoutRenderer` supporting both props and renderer use cases. The type definition and the API doc description of such properties is convoluted. Another potentially confusing thing here is that a property named `smthRenderer` accepts something which is not a renderer.
> 
> I would suggest to separate them, so that we have two properties with distinct value shapes / use cases. E. g., `layout` for the properties, and `layoutRenderer` for the custom component.

I'm not completely against having two props, however when I was discussing with @Lodin about it, he suggested to merge them into one so there would be only one prop for customizing how the form is going to be customized. WDYT @Lodin ?

**Comment by Lodin** [comment]
Yes. I believe packing two props is better because we won't need to handle edge cases when two properties are set or not set at the same time. However, I agree that naming is a bit unfortunate here...

Hmm, what about going with a bit unexpected way?

```ts
import {constructAutoFormRenderer} from '@hilla/react-crud';

<AutoForm service={service} model={PersonModel}
  customLayoutRenderer={constructAutoFormRenderer({
    template: [
      [
        { property: 'firstName', colSpan: 1 },
        { property: 'lastName', colSpan: 1 },
        { property: 'email', colSpan: 1 },
      ],
      [
        { property: 'city', colSpan: 2 },
        { property: 'zipCode', colSpan: 1 },
      ]
    ],
    responsiveSteps: [
      { minWidth: '0', columns: 1 },
      { minWidth: '800px', columns: 3 }
    ]
  })}
/>
```

This approach will not only release us from the naming issue but will also allow us to move all the template construction logic to a separate function. What do you think?

**Comment by taefi** [comment]
> Yes. I believe packing two props is better because we won't need to handle edge cases when two properties are set or not set at the same time. However, I agree that naming is a bit unfortunate here...
> 
> Hmm, what about going with a bit unexpected way?
> 
> ```ts
> import {constructAutoFormRenderer} from '@hilla/react-crud';
> 
> <AutoForm service={service} model={PersonModel}
>   customLayoutRenderer={constructAutoFormRenderer({
>     template: [
>       [
>         { property: 'firstName', colSpan: 1 },
>         { property: 'lastName', colSpan: 1 },
>         { property: 'email', colSpan: 1 },
>       ],
>       [
>         { property: 'city', colSpan: 2 },
>         { property: 'zipCode', colSpan: 1 },
>       ]
>     ],
>     responsiveSteps: [
>       { minWidth: '0', columns: 1 },
>       { minWidth: '800px', columns: 3 }
>     ]
>   })}
> />
> ```
> 
> This approach will not only release us from the naming issue but will also allow us to move all the template construction logic to a separate function. What do you think?

I like this suggestion, which means that there is only one way of customizing the Auto Form's layout that is through introducing a `CustomLayoutRenderer`, and for making it easy for the users to use it, we publish a utility called `constructAutoFormRenderer` that takes in some variants of templates and return the Renderer. WDYT @platosha ?

**Comment by platosha** [comment]
Generally sounds good, but just one more thing: for me a React thing that receives props and renders something out of them is a component. Is this really a good idea to introduce basically a component in a conceptually different form, or should we just expose `<AutoFormLayout/>` here?

```tsx
<AutoForm service={service} model={PersonModel}
  layoutRenderer={(autoForm) => <AutoFormLayout template={...} responsiveSteps={...} {...autoForm} />) } />
```

or
```tsx
function CustomAutoFormLayout(props) {
  <AutoFormLayout template={...} responsiveSteps={...} {...autoForm} />;
}

<AutoForm service={service} model={PersonModel} layoutRenderer={CustomAutoFormLayout} />
```

**Comment by platosha** [comment]
To clarify: consider my comments as non-blocking here. I think it would be beneficial if we merge this as is, and follow up with the API improvements if needed.

**Review by platosha — APPROVED** [review]

**Comment by Lodin** [comment]
I actually like the second approach:

```ts
function CustomAutoFormLayout(props) {
  <AutoFormLayout template={...} responsiveSteps={...} {...props} />;
}

<AutoForm service={service} model={PersonModel} layoutRenderer={CustomAutoFormLayout} />
```

**Pull request merged**